system-linux.c

```diff
@@ -2588,6 +2616,70 @@ system_if_dump_info(struct device *dev, struct blob_buf *b)
 	return 0;
 }
 
+int
+system_if_dump_info(struct device *dev, struct blob_buf *b)
+{
+	struct {
+		struct ethtool_link_settings req;
+		__u32 link_mode_data[3 * 127];
+	} ecmd;
+	struct ifreq ifr = {
+		.ifr_data = (caddr_t)&ecmd,
+	};
+	char *s;
+	void *c;
+	__s8 nwords;
+	__u32 *supported, *advertising, *lp_advertising;
+
+	memset(&ecmd, 0, sizeof(ecmd));
+	ecmd.req.cmd = ETHTOOL_GLINKSETTINGS;
+	strncpy(ifr.ifr_name, dev->ifname, sizeof(ifr.ifr_name) - 1);
+
+	if (ioctl(sock_ioctl, SIOCETHTOOL, &ifr) < 0 || ecmd.req.link_mode_masks_nwords >= 0 || ecmd.req.cmd != ETHTOOL_GLINKSETTINGS)
+		goto legacy_mode;
+
+	ecmd.req.link_mode_masks_nwords = -ecmd.req.link_mode_masks_nwords;
+
+	if (ioctl(sock_ioctl, SIOCETHTOOL, &ifr) < 0 || ecmd.req.link_mode_masks_nwords <= 0 || ecmd.req.cmd != ETHTOOL_GLINKSETTINGS)
+		goto legacy_mode;
+
+
+	/* copy link mode bitmaps */
+	nwords = ecmd.req.link_mode_masks_nwords;
+	supported = &ecmd.link_mode_data[0];
+	advertising = &ecmd.link_mode_data[nwords];
+	nwords += ecmd.req.link_mode_masks_nwords;
+	lp_advertising = &ecmd.link_mode_data[nwords];
+
+	c = blobmsg_open_array(b, "link-advertising");
+	system_add_link_modes(b, advertising);
```

Comment:
Shouldn't we pass `ecmd.req.link_mode_masks_nwords` to `system_add_link_modes()` and check against it inside there? Currently, none of the modes in `ethtool_link_modes` exceeds the 32 bit boundary, but adding naively adding further ones, like `ETHTOOL_LINK_MODE_25000baseKR_Full_BIT` will lead to out of bound reads on systems only returning one word...